packages/ngtools/webpack/src/ngcc_processor.ts

```diff
@@ -0,0 +1,103 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import { JsonParseMode, parseJson } from '@angular-devkit/core';
+import * as ts from 'typescript';
+import { time, timeEnd } from './benchmark';
+import { WebpackCompilerHost } from './compiler_host'; // tslint:disable-line:no-circular-imports
+import { workaroundResolve } from './utils';
+
+export class NgccProcessor {
+  private _ngcc: typeof import('@angular/compiler-cli/ngcc') | undefined;
+  private _processedModules = new Set<string>();
+
+  constructor(
+    private readonly propertiesToConsider: string[],
+    private readonly host: WebpackCompilerHost,
+  ) {
+    // this is done for the sole reason that @ngtools/webpack
+    // support versions of Angular that don't have NGCC API
+    try {
+      this._ngcc = require('@angular/compiler-cli/ngcc');
+    } catch {
+    }
+  }
+
+  processModule(
+    moduleName: string,
+    resolvedModule: ts.ResolvedModule | ts.ResolvedTypeReferenceDirective,
+  ): void {
+    const resolvedFileName = resolvedModule.resolvedFileName;
+    if (
+      !this._ngcc
+      || !resolvedFileName
+      || moduleName.startsWith('.')
+      || this._processedModules.has(moduleName)) {
+      // Skip when module is unknown, relative or NGCC compiler is not found or already processed.
+      return;
+    }
+
+    const packageJsonPath = this.tryResolvePackage(moduleName, resolvedFileName);
+    if (!packageJsonPath) {
+      // add it to processed to the second time round we skip this
+      this._processedModules.add(moduleName);
+
+      return;
+    }
+
+    const packageJsonData = this.host.readFile(packageJsonPath) as string;
+    const packageJson
+      = parseJson(packageJsonData, JsonParseMode.Loose) as Record<string, unknown> | null;
+    if (!packageJson) {
+      throw new Error(`Cannot parse '${packageJsonPath}' as JSON.`);
+    }
+
+    const format = this.propertiesToConsider.find(mf => typeof packageJson[mf] === 'string');
+    if (!format || this._ngcc.hasBeenProcessed(packageJson, format)) {
+      // Unknown format or module has already been processed by NGCC.
+      this._processedModules.add(moduleName);
+
+      return;
+    }
```

Comment:
If you are going to have to load the packageJson anyway then there is not much performance benefit in not just calling ngcc everytime. Ngcc will still check the build markers and also noop if the format is not recognized.

So you could simplify this whole block to just call ngcc with the `packageJsonPath`.